### PR TITLE
[VideoLibrary] Do not scan unsupported plugin sources + fix database cleaning

### PR DESCRIPTION
**Pull request by AkariDN**
Continue #13566:
* Implemented protection: plugins without media library scanning support will not be scanned and refreshed even if they are used as video source
* Fix video library database cleaning for plugin sources
* Other minor fixes

### [VideoLibrary] Do not scan and refresh unsupported plugin sources
Video library scanning and refreshing is allowed only for plugin URLs that indicated in addon.xml (within extension point "xbmc.python.pluginsource"). These URLs are root paths for specified content. If no URLs are specified, then scanning is completely disabled for such plugin. Different URLs with the same content can be specified multiple times.
```XML
    <extension point="xbmc.python.pluginsource" library="default.py">
        <provides>video</provides>
        <medialibraryscanpath content="movies">movies/</medialibraryscanpath>
        <medialibraryscanpath content="tvshows">tvshows/</medialibraryscanpath>
    </extension>
```
Currently content attribute can be "movies", "tvshows" or "musicvideos".
Path (URL) can be absolute or relative.

```
In the addon.xml example above at following pathes:
plugin://my.favorite.plugin/ - scanning and refreshing is forbidden for any content
plugin://my.favorite.plugin/?mode=movies - forbidden for any content
plugin://my.favorite.plugin/play/movie/123/ - forbidden for any content
plugin://my.favorite.plugin/movies/ - allowed only for movies
plugin://my.favorite.plugin/movies/12345/ - allowed only for movies
plugin://my.favorite.plugin/tvshows/?id=12345 - allowed only for tvshows
plugin://my.favorite.plugin/tvshows/play/?episode_id=777 - allowed only for tvshows
```
In practice, it's desirable that addon developers provide a special plugin url (node), which is intended for scanning, and specify it in addon.xml, so that scanning can be performed only on this special url. For example:
* plugin://my.favorite.plugin/ - standard plugin operation mode
* plugin://my.favorite.plugin/library/ - this path is used only for library scanning and must be added to addon.xml

Note: this is a breaking change. Scanning unsupported plugins is disabled by default. All plugins with media library scanning support must have corresponding entries in addon.xml.

### [VideoLibrary] Fix database cleaning for plugin sources
Applies only to plugins that support media library scanning. In order to check existance of some plugin URL, video database cleaner calls plugin at specified path with option "kodi_action=check_exists" and examines result. Plugin must check whether specified path exists (e.g. using cache) and return result with setResolvedUrl(). Simplified example:
```Python
if args['kodi_action'] == 'check_exists':
    xbmcplugin.setResolvedUrl(handle, check(path), xbmcgui.ListItem())
```
Plugins must correctly handle this mode - just check the existence and don't play video, open dialogs or something else.

### [VideoLibrary] Other minor fixes
* Correctly handle empty hash for tvshow folder (i.e. no episodes) calculated by plugin
* Case insensitive compare for tvshow folder hashes
* Optimize plugin root check in InvalidatePathHash

## Motivation and Context
As tamland said:
> Since plugins are allowed to do anything this will cause all kind of weird and possibly dangerous side-effects... And it's absolutely dangerous because for starters plugins can be unbound, they can open dialog, start playback and whatever as a side-effect of opening directories; and many do. There is no way a user can know this.

## How Has This Been Tested?
* Build tested on Windows and Linux
* Functionality tested with special [test plugin](https://github.com/AkariDN/plugin.video.testlib)

## Types of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Cosmetic change (non-breaking change that doesn't touch code)
- [ ] None of the above (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/video/VideoDatabase.h [hunk 1]
```diff
@@ -655,9 +655,10 @@ class CVideoDatabase : public CDatabase
    we've found a scraper directly (shows).  Any folders inbetween are treated as seasons (regardless of whether
    they actually are seasons). Note that any subfolders in movies will be treated as movies.
    \param strPath path to start searching in.
+   \param checkTvShows check for episodes/seasons in the path in case of tvshows.
    \return A content type string for the current path.
    */
-  std::string GetContentForPath(const std::string& strPath);
+  std::string GetContentForPath(const std::string& strPath, bool checkTvShows = true);
```
Comment (FernetMenta) [inline xbmc/video/VideoDatabase.h]: no default values for parameters parameters

**Comment by AkariDN** [comment]
Thank you. Fixed.

File: xbmc/filesystem/PluginDirectory.cpp [hunk 2]
```diff
@@ -604,3 +606,46 @@ float CPluginDirectory::GetProgress() const
     return (m_listItems->Size() * 100.0f) / m_totalItems;
   return 0.0f;
 }
+
+bool CPluginDirectory::IsMediaLibraryScanningAllowed(const std::string content, const std::string& strPath)
```
Comment (notspiff) [inline xbmc/filesystem/PluginDirectory.cpp]: pass by ref

File: xbmc/filesystem/PluginDirectory.cpp [hunk 3]
```diff
@@ -604,3 +606,46 @@ float CPluginDirectory::GetProgress() const
...
+  if (it == paths.end())
+    return false;
+  std::string path = url.GetFileName();
+  for (const auto& p : it->second)
+    if (p.empty() || p == "/" || URIUtils::PathHasParent(path, p))
+      return true;
+  return false;
+}
+
+bool CPluginDirectory::CheckExists(const std::string content, const std::string& strPath)
```
Comment (notspiff) [inline xbmc/filesystem/PluginDirectory.cpp]: pass by ref

**Review by notspiff — CHANGES_REQUESTED** [review]

**Comment by AkariDN** [comment]
Thank you. Done.

**Comment by DaveTBlake** [comment]
Just wondering why the switch to using m_pDS2 in `CVideoDatabase::CleanDatabase`? I thought that the video lib job queue prevented any other changes during cleanup, so thought that using m_pDS1 was fine.

**Comment by Rechi** [comment]
@AkariDN the fixup commit has to be squashed, we want a clean git history.

**Comment by AkariDN** [comment]
> why the switch to using m_pDS2

GetScraperForPath() that is called from (!m_pDS->eof()) loop uses the same dataset m_pDS.
> the fixup commit has to be squashed

Done

**Review by notspiff — APPROVED** [review]

**Review by DaveTBlake — APPROVED** [review]
Thanks for clarifying

**Pull request merged**